core/authentication_api.php

```diff
@@ -541,6 +541,30 @@ function auth_generate_random_password() {
 }
 
 /**
+ * Generate a random string to use when we should not store the actual password.
+ * When relying on an external authentication system (e.g. LDAP), we should not
+ * store the user's password for security reasons. To avoid creating another
+ * security hole by populating the password field with a known value, we use a
+ * random string that can be identified as a 'nocache' password by its prefix.
+ * @return string Random password, prefixed by '!'
+ * @access public
+ */
+function auth_generate_nocache_password() {
+	return '!' . crypto_generate_strong_random_string( 31 );
```

Comment:
From memory even though you supply 31, it will still generate a 32 character since it goes to the next number that is divisible by 4.  I would generate the 32 and then truncate to 31 +  a comment to explain.

Reply:
Shouldn't this be using crypto_generate_uri_safe_nonce() instead of crypto_generate_strong_random_string()? I had database issues with the output of the latter.